### PR TITLE
Retrieve launch plan interfaces from admin

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
When Propeller comes across a launch plan node in a `DynamicJobSpec` it will hit Admin to retrieve the interface for the LP to do the interface check.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Please see the issue linked below and also the SDK PR https://github.com/lyft/flytekit/pull/92 for more information.  A sample dynamic job spec object has been uploaded here as well.  Please see the text file for the type of dynamic job spec this PR is meant to support.
[dynamic_job_spec.txt](https://github.com/lyft/flytepropeller/files/4430252/dynamic_job_spec.txt)

* Added a `GetLaunchPlan` function to a `launchplan/Reader` interface which sits alongside the `launchplan/Executor` interface.  Admin client wrapper now satisfies both interfaces.
* Added a call to that function in the dynamic job handler `buildContextualDynamicWorkflow` function.

## Tracking Issue
https://github.com/lyft/flyte/issues/139

## Follow-up issue
https://github.com/lyft/flyte/issues/246
This PR will be deprecated upon completion of this issue.

### CODE REVIEW

**Comment by kumare3** [comment]
What is this for? Dynamic LP?

**Comment by wild-endeavor** [comment]
Yes.

File: pkg/controller/nodes/dynamic/handler.go [hunk 1]
```diff
@@ -377,8 +379,19 @@ func (d dynamicNodeTaskNodeHandler) buildContextualDynamicWorkflow(ctx context.C
...
+	requirements, err := compiler.GetRequirements(wf, djSpec.Subworkflows)
+	if err != nil {
+		return nil, true, err
+	}
+
+	launchPlanInterfaces, err := d.getLaunchPlanInterfaces(ctx, requirements.GetRequiredLaunchPlanIds())
+	if err != nil {
+		return nil, true, err
+	}
+
```
Comment (EngHabu) [inline pkg/controller/nodes/dynamic/handler.go:393]: Mind adding a TODO to get the tasks we don't already have from Admin as well...

File: pkg/compiler/admin.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package compiler
+
+import (
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+)
+
+// Should we call this something else? It can be used more generally if ever necessary.
```
Comment (EngHabu) [inline pkg/compiler/admin.go:8]: Historically: I wanted to make both tasks and launchplan  parameters in the compiler to take InterfaceProviders... but then launch plans diverged  (with expected* stuff) maybe there is still hope?

Comment (wild-endeavor) [reply pkg/compiler/admin.go]: I can't think of a good name... InterfaceProviderProvider was the best I could come up with, so I'm leaving it as is for now.

File: pkg/controller/nodes/dynamic/handler.go (hunk 1 above)
Comment (wild-endeavor) [reply pkg/controller/nodes/dynamic/handler.go:393]: Yup.

File: pkg/controller/nodes/subworkflow/launchplan/launchplan.go [hunk 3]
```diff
@@ -31,6 +31,9 @@ type Executor interface {
 	// Kill a remote execution
 	Kill(ctx context.Context, executionID *core.WorkflowExecutionIdentifier, reason string) error
 
+	// Get the definition of a launch plan. This is primarily used to ensure all the TypedInterfaces match up before actually executing.
+	GetLaunchPlan(ctx context.Context, launchPlanRef *core.Identifier) (*admin.LaunchPlan, error)
```
Comment (kumare3) [inline pkg/controller/nodes/subworkflow/launchplan/launchplan.go:35]: I think we should create a separate interface for this. And then pass that interface to Dynamic task. This is weird coupling - subworkflow only ever launches a launchplan, having this method is confusing and in the future may lead to weird uses. But having a separate interface will prevent these problems, and also make uses cases clear

File: pkg/controller/nodes/subworkflow/launchplan/admin.go [hunk 4]
```diff
@@ -97,6 +97,26 @@ func (a *adminLaunchPlanExecutor) GetStatus(ctx context.Context, executionID *co
 	return item.ExecutionClosure, item.SyncError
 }
 
+func (a *adminLaunchPlanExecutor) GetLaunchPlan(ctx context.Context, launchPlanRef *core.Identifier) (*admin.LaunchPlan, error) {
```
Comment (kumare3) [inline pkg/controller/nodes/subworkflow/launchplan/admin.go:100]: Having this method here is fine, but having a separate interface for this method is useful to differentiate the use cases and usage paths

File: pkg/compiler/admin.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package compiler
+
+import (
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+)
+
+// This object is meant to satisfy github.com/lyft/flytepropeller/pkg/compiler/common.InterfaceProvider
```
Comment (kumare3) [inline pkg/compiler/admin.go:8]: This comment does not answer why this is needed here?

**Comment by kumare3** [comment]
Thank you for the changes.

File: pkg/compiler/admin.go (hunk 5 above)
Comment (wild-endeavor) [reply pkg/compiler/admin.go:8]: We need a struct to satisfy the interface because the function that we pass it to, `compiler.CompileWorkflow` takes the `common.InterfaceProvider` interface.
